### PR TITLE
Prevent changing the working time of student exams after the exam is visible

### PR DESCRIPTION
**Pull request by krusche**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context
Changing the working time after the exam is visible, leads to issues with lock and unlock programming exercise operations. This can also be confusing during testing

### Description
Instructors can now only change the working time of an individual student before the exam is visible

### Steps for Testing
1. Generate student exams
2. Change the working time of one student **before** the exam is visible --> should work
3. Try to change the working time of one student **after** the exam is visible --> should **not** work

### CODE REVIEW

File: src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts [hunk 1]
```diff
@@ -107,8 +108,29 @@ export class StudentExamDetailComponent implements OnInit {
...
+                Validators.required,
+            ]),
         });
     }
+
+    examIsVisible(): boolean {
+        if (this.studentExam.exam) {
+            return moment(this.studentExam.exam.visibleDate).isBefore(moment());
+        }
+        // if the exam is not visible, we rather disable the form
```
Comment (sleiss) [inline src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts:127]:
```suggestion
        // if the exam is not loaded, we rather disable the form
```

File: src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts [hunk 2]
```diff
@@ -107,8 +108,29 @@ export class StudentExamDetailComponent implements OnInit {
...
+                Validators.min(0),
+                Validators.max(59),
+                Validators.required,
+            ]),
         });
     }
+
+    examIsVisible(): boolean {
+        if (this.studentExam.exam) {
+            return moment(this.studentExam.exam.visibleDate).isBefore(moment());
```
Comment (sascha11110) [inline src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts:125]: You may want to add a comment here to make this more clear:
```suggestion
            // Disable the form to edit the working time if the exam is already visible
            return moment(this.studentExam.exam.visibleDate).isBefore(moment());
```

File: src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts [hunk 3]
```diff
@@ -107,8 +108,29 @@ export class StudentExamDetailComponent implements OnInit {
...
+        }
+        // if the exam is not visible to students yet, we disable the form to edit the working time
```
Comment (sascha11110) [inline src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts:127]: I don't get this comment. For two reasons:
- The purpose of the PR is to ensure that the working time cannot be changed after the exam is visible to the students. If there is no visibleDate the working time can be changed --> so the method should return false (--> form is editable)
- The first part (*if the exam is not visible to students yet*) is a little confusing. If the visibleDate is not set would make more sense here

File: src/test/java/de/tum/in/www1/artemis/StudentExamIntegrationTest.java [hunk 4]
```diff
@@ -70,7 +70,13 @@ public void initTestCase() {
     }
 
     @AfterEach
-    public void resetDatabase() {
+    public void resetDatabase() throws Exception {
+
+        // change back to instructor user
+        database.changeUser("instructor1");
+        // Clean up to prevent exceptions during reset database
+        request.delete("/api/courses/" + course1.getId() + "/exams/" + exam1.getId(), HttpStatus.OK);
```
Comment (sascha11110) [inline src/test/java/de/tum/in/www1/artemis/StudentExamIntegrationTest.java:78]: We should find a better solution for this soon :-D

Comment (julian-christl) [reply src/test/java/de/tum/in/www1/artemis/StudentExamIntegrationTest.java:78]: In my project I always use a try{}finally{} and within the finally I delete all of the entities created by accessing the repository directly
if(repo.existsById(...)){
    repo.deleteById(...);
}
Something like that. Is that something that might help you?

Comment (krusche) [reply src/test/java/de/tum/in/www1/artemis/StudentExamIntegrationTest.java:78]: we have a better solution already, I added a cleanup database service on `develop` so this line might not even be necessary, but still a good test that the DELETE Rest call does work as intended

**Comment by sascha11110** [comment]
@julian-christl I think waiting for the exam to become visible with an open student exam page is a very rare use case. I think we can ignore this small client irritation. As long as the server does not allow to 

I've also found the problem with the tooltip. I think this is rather complicated to fix as the div with the tooltip will be applied to the element in focus. Our disabled buttons have an `opacity` property set that will be applied to all child elements too. So we cannot fix it with a simple CSS rule. So I suggest to leave it as it is.

**Comment by krusche** [comment]
> Working as intended.
> However I do have one small issue which isn't breaking however:
> 
> If you open the student page before the exam is visible and wait for the exam becoming visible, the Save button turns inactive. So far intended behaviour. However the inputs are still editable (even though you can't click the save button). Once you reload the inputs become disabled.
> 
> And another thing, style issue: Tooltops become opaque if the e.g. button they are applied to are opaque as well.

Thanks for testing, I can live with these two issues. Tbh: I am not an expert in Angular and Bootstrap and there are probably other people who can implement this in a better way. For now we actually want to prevent confusion, also with respect to #1804

File: src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts (hunk 3 above)
Comment (krusche) [reply src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts:127]: the comment is wrong, I'll update it. This should just prevent an edge case and type script error when for some reason the studentExam.exam is not loaded or undefined

Comment (krusche) [inline src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts:127]:
```suggestion
        // if exam is undefined, the form to edit the working time is disabled
```

**Pull request merged**